conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java

```diff
@@ -115,53 +126,301 @@ private BeanBuilderGenerator(
+        specBuilder.addAnnotation(AnnotationSpec.builder(JsonDeserialize.class)
+                .addMember(
+                        "builder",
+                        "$T.class",
+                        ClassName.get(objectClass.packageName(), objectClass.simpleName(), implementationClassName))
+                .build());
+    }
+
+    private void addStageInterfacesAndBuilderMethod(
+            TypeSpec.Builder specBuilder,
+            List<EnrichedField> allFields,
+            List<EnrichedField> fieldsNeedingBuilderStage) {
+        Preconditions.checkState(
+                !fieldsNeedingBuilderStage.isEmpty(), "Expected at least one field needing a builder stage.");
+        List<TypeSpec> interfaces = generateStageInterfaces(
+                objectClass,
+                builderClass,
+                typeMapper,
+                fieldsNeedingBuilderStage,
+                allFields.stream()
+                        .filter(f -> !fieldsNeedingBuilderStage.contains(f))
+                        .collect(ImmutableList.toImmutableList()),
+                safetyEvaluator);
+
+        List<TypeName> interfaceClassNames = interfaces.stream()
+                .map(stageInterface ->
+                        ClassName.get(objectClass.packageName(), objectClass.simpleName(), stageInterface.name))
+                .map(Primitives::box)
+                .collect(Collectors.toList());
+
+        TypeSpec builderInterface = TypeSpec.interfaceBuilder(STAGED_BUILDER_INTERFACE_NAME)
+                .addModifiers(Modifier.PUBLIC)
+                .addMethods(interfaces.stream()
+                        .map(stageInterface -> stageInterface.methodSpecs)
+                        .flatMap(List::stream)
+                        .map(method -> method.toBuilder()
+                                .addAnnotation(Override.class)
+                                .returns(
+                                        method.name.equals("build")
+                                                ? method.returnType
+                                                : ClassName.get(
+                                                        objectClass.packageName(),
+                                                        objectClass.simpleName(),
+                                                        STAGED_BUILDER_INTERFACE_NAME))
+                                .build())
+                        .collect(Collectors.toList()))
+                .addSuperinterfaces(interfaceClassNames)
+                .build();
+
+        specBuilder
+                .addTypes(interfaces)
+                .addType(builderInterface)
+                .addMethod(MethodSpec.methodBuilder("builder")
+                        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                        .returns(interfaceClassNames.get(0))
+                        .addStatement(String.format("return new %s()", STAGED_BUILDER_IMPLEMENTATION_NAME))
```

Comment:
Javapoet supports templating along these lines:
```suggestion
                        .addStatement("return new $N()", STAGED_BUILDER_IMPLEMENTATION_NAME)
```

Docs in the readme: https://github.com/square/javapoet#javapoet

Reply:
nice!

Reply:
Nice!